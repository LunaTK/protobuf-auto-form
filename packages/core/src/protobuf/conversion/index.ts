export * from './proto2Form';
export * from './form2Proto';
