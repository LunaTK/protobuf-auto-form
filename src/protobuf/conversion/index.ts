export * from './protoObj2Form'
export * from './form2ProtoObj'
