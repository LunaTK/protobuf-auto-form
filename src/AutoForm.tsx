import React, { useMemo } from 'react';
import protobuf from 'protobufjs';
import { FormProvider, useForm, UseFormReturn } from 'react-hook-form';
import './index.css';
import Message from './protobuf/input/Message';
import ErrorAlert from './common/ErrorAlert';
import { finalize, protoObjToForm } from './protobuf/conversion';
import { AutoFormContext, AutoFormProvider } from './context';

interface AutoFormProps<T = any> extends React.HTMLAttributes<HTMLFormElement> {
  namespace: protobuf.Namespace
  messageType: string
  form?: UseFormReturn
  initialState?: T
  onSubmitValid?: (values: T) => void
  hideFieldType?: AutoFormContext['hideFieldType']
  camelCaseLabel?: AutoFormContext['camelCaseLabel']
  wellKnownFields?: AutoFormContext['wellKnownFields']
  wellKnownTypes?: AutoFormContext['wellKnownTypes']
}

const AutoForm = <T, >(props: AutoFormProps<T>) => {
  const {
    namespace,
    messageType,
    children,
    onSubmitValid,
    initialState,
    hideFieldType = false,
    camelCaseLabel = true,
    wellKnownFields = {},
    wellKnownTypes = {},
    ...rest
  } = props;
  const reflectionObj = useMemo(() => {
    try {
      return namespace.resolveAll().lookupType(messageType);
    } catch (e) {
      return null;
    }
  }, [namespace, messageType]);
  const defaultValues = useMemo(() => {
    if (!initialState || !reflectionObj) return undefined;

    const formState = protoObjToForm(initialState, reflectionObj);
    return formState as Record<string, {}>;
  }, [initialState, reflectionObj]);

  const methods = useForm({ defaultValues });

  if (!reflectionObj) {
    return <ErrorAlert>{`Cannot find message type: ${messageType}`}</ErrorAlert>;
  }

  return (
    <FormProvider {...methods}>
      <AutoFormProvider value={{
        hideFieldType, camelCaseLabel, wellKnownFields, wellKnownTypes,
      }}
      >
        <form
          {...rest}
          onSubmit={methods.handleSubmit((values) => {
            onSubmitValid?.(finalize(values, reflectionObj));
          })}
        >
          <Message type={reflectionObj} />
          {children}
        </form>
      </AutoFormProvider>
    </FormProvider>
  );
};

export default AutoForm;
