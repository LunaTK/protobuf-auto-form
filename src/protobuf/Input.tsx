import React from 'react';
import protobuf from 'protobufjs';
import RepeatedInput from './input/Repeated';
import BasicInput, { isBasicType } from './input/Basic';
import EnumInput from './input/Enum';
import Message from './input/Message';

interface InputProps {
  field: protobuf.Field
  name: string
  ignoreRepeatAndMap?: boolean
}

const Input: React.FC<InputProps> = ({ field, name, ignoreRepeatAndMap }) => {
  const {
    resolvedType, type, repeated,
  } = field;

  if (!ignoreRepeatAndMap && repeated) {
    return <RepeatedInput field={field} name={name} />;
  } if (isBasicType(type)) {
    return <BasicInput name={name} type={type} />;
  } if (resolvedType instanceof protobuf.Enum) {
    return <EnumInput name={name} type={resolvedType} />;
  } if (resolvedType instanceof protobuf.Type) {
    return <Message name={name} type={resolvedType} />;
  }

  throw new Error(`Unresolved type "${type}" from field "${name}"`);
};

export default Input;
