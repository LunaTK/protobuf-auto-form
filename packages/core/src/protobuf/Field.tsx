import React from 'react';
import protobuf from 'protobufjs';
import { pascalCase } from 'change-case';
import OneofField, { isProto3Optional } from './OneofField';
import Input from './input/Input';
import { useAutoFormCtx } from '../context';
import { FieldOptions } from '../models';
import withPrependAppend from '../hoc/withPrependAppend';
import AutoFormLabel from '../AutoFormLabel';
import { join } from '../utils';

interface Props {
  field: protobuf.Field | protobuf.OneOf;
  parentName: string;
  hideLabel?: boolean;
  options?: FieldOptions;
}

const toSpaceSeperated = (name: string) =>
  pascalCase(name).replace(/([A-Z][a-z])/g, ' $1').replace(/(\d)/g, ' $1');

const getTypeLabel = (field: Props['field']) => {
  if (field instanceof protobuf.OneOf) {
    return isProto3Optional(field) ? 'optional' : 'oneof';
  }
  if (field instanceof protobuf.MapField) {
    return `map<${field.keyType}, ${field.type}>`;
  }
  if (field.repeated) {
    return `repeated<${field.type}>`;
  }
  return field.type;
};

const Field: React.FC<Props> = ({
  field,
  parentName,
  hideLabel = false,
  options,
}) => {
  const { camelCaseLabel, hideFieldType } = useAutoFormCtx();
  if (options?.hidden === true) return null;

  const isFlattenedMessage = 'resolvedType' in field && options?.flatten;
  const ignoreLabel = hideLabel || isFlattenedMessage;

  return (
    <>
      {!ignoreLabel && (
        <AutoFormLabel
          typeLabel={!hideFieldType && getTypeLabel(field)}
          required={!!options?.rules}
        >
          {options?.label ||
            (camelCaseLabel ? field.name : toSpaceSeperated(field.name))}
        </AutoFormLabel>
      )}

      {field instanceof protobuf.OneOf ? (
        <OneofField parentName={parentName} oneof={field} options={options} />
      ) : (
        <Input
          name={join(parentName, field.name)}
          field={field}
          options={options}
        />
      )}
    </>
  );
};

export default withPrependAppend(Field);
