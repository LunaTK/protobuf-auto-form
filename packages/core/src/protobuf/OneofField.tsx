import React, { ReactNode } from 'react';
import protobuf from 'protobufjs';
import { useFormContext } from 'react-hook-form';
import RadioButton from '../common/RadioButton';
import Input from './input/Input';
import { FieldOptions } from '../models';
import { useChildFields } from '../hooks';
import { join, toSpaceSeperated } from '../utils';
import { useAutoFormCtx } from '../context';

interface OneofProps {
  parentName: string;
  oneof: protobuf.OneOf;
  options?: FieldOptions;
}

const OPT_UNSET = '__unset__';

const OneofElement: React.FC<{ flatten?: boolean; children: ReactNode }> = ({
  flatten,
  children,
}) => {
  if (flatten) return <>{children}</>;
  return <div className="my-2 not-first-desc:ml-8">{children}</div>;
};

export const isProto3Optional = (oneof: protobuf.OneOf) =>
  oneof.fieldsArray[0].options?.proto3_optional;

const OneofField: React.FC<OneofProps> = ({ parentName, oneof, options }) => {
  const { fieldOptions } = useChildFields(options);
  const { camelCaseLabel } = useAutoFormCtx();
  const { watch, getValues, setValue, register } = useFormContext();
  const oneofFullName = join(parentName, oneof.name);
  if (!getValues(oneofFullName)) {
    setValue(oneofFullName, OPT_UNSET);
  }
  const selected: string = watch(oneofFullName);
  const renderLabel = (f: protobuf.Field) =>
    fieldOptions[f.name]?.label ??
    (camelCaseLabel ? f.name : toSpaceSeperated(f.name));

  const getDropdownContent = () => {
    const selectedField = oneof.fieldsArray.find((f) => f.name === selected);
    if (!selectedField) return null;

    return (
      <>
        <select
          {...register(oneofFullName)}
          className="max-w-xs mb-2 select select-bordered select-sm"
        >
          {oneof.fieldsArray.map((f) => (
            <option key={f.name} value={f.name}>
              {renderLabel(f)}
            </option>
          ))}
          {isProto3Optional(oneof) && <option value={OPT_UNSET}>None</option>}
        </select>

        <Input
          name={join(parentName, selectedField.name)}
          field={selectedField}
          options={fieldOptions[selectedField.name]}
        />
      </>
    );
  };

  const getRadioContent = () => {
    return (
      <>
        {oneof.fieldsArray.map((f) => (
          <OneofElement key={f.name} flatten={options?.flatten}>
            <RadioButton
              value={f.name}
              name={oneofFullName}
              label={renderLabel(f)}
            />
            {selected === f.name && (
              <Input
                name={join(parentName, f.name)}
                field={f}
                options={fieldOptions[f.name]}
              />
            )}
          </OneofElement>
        ))}
        {isProto3Optional(oneof) && (
          <OneofElement>
            <RadioButton name={oneofFullName} label="None" value={OPT_UNSET} />
          </OneofElement>
        )}
      </>
    );
  };

  const content = options?.dropdown ? getDropdownContent() : getRadioContent();

  if (options?.flatten) {
    return content;
  }

  return <div>{content}</div>;
};

export default OneofField;
