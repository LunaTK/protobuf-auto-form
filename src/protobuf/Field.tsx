import React from 'react';
import protobuf from 'protobufjs';
import OneofInput from './input/Oneof';
import Input from './Input';

interface Props {
  field: protobuf.Field | protobuf.OneOf
  parentName: string
}

const getTypeLabel = (field: Props['field']) => {
  if (field instanceof protobuf.OneOf) {
    return 'oneof';
  } if (field instanceof protobuf.MapField) {
    return `map<${field.keyType}, ${field.type}>`;
  }
  return field.type;
};

const Label: React.FC<{
  field: Props['field']
}> = ({ field }) => (
  <span className="text-right inline-flex flex-col">
    <span className="leading-tight font-bold">
      {field.name}
    </span>
    <span className="text-slate-400 text-sm">
      {getTypeLabel(field)}
    </span>
  </span>
);

const Field: React.FC<Props> = ({ field, parentName }) => (
  <>
    <Label field={field} />
    {field instanceof protobuf.OneOf
      ? <OneofInput parentName={parentName} oneof={field} />
      : <Input name={parentName ? `${parentName}.${field.name}` : field.name} field={field} />}
  </>
);

export default Field;
