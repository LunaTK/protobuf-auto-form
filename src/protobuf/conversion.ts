import protobuf from 'protobufjs';
import { AutoFormContext } from '../context';

const isUnselectedOneofField = (data: any, field: protobuf.Field) => {
  if (!field.partOf) return false;
  return data[field.partOf.name] !== field.name;
};

const pruneUnselectedOneofValues = (_data: any, type: protobuf.Type, isRepeated: boolean = false) => {
  if (!_data) return _data;
  if (isRepeated) {
    return _data.map((element: any) => pruneUnselectedOneofValues(element, type));
  }

  const data = JSON.parse(JSON.stringify(_data));

  type.fieldsArray.forEach((field) => {
    if (isUnselectedOneofField(data, field)) {
      data[field.name] = undefined;
      return;
    }
    if (field.resolvedType instanceof protobuf.Type) {
      data[field.name] = pruneUnselectedOneofValues(data[field.name], field.resolvedType, field.repeated);
    }
  });

  return data;
};

const formToProtoObj = (context: AutoFormContext) => {
  const encode = (formState: unknown, type: protobuf.Type | protobuf.Enum | null) => {
    if (formState === undefined) return undefined;
    if (!(type instanceof protobuf.Type)) return formState;
    const ret = { ...(formState as Record<string, unknown>) };
  
    type.fieldsArray.forEach((f) => {
      if (f.repeated) {
        ret[f.name] = (ret[f.name] as Record<string, unknown>[])?.map(({ $value }) =>
        encode($value, f.resolvedType),
        );
      } else if (f.map) {
        ret[f.name] = Object.fromEntries(
          (ret[f.name] as Record<string, unknown>[])?.map(({ $key, $value }) => [
            $key,
            encode($value, f.resolvedType),
          ]),
        );
      } else if (isUnselectedOneofField(ret, f)) {
        ret[f.name] = undefined;
      } else {
        ret[f.name] = encode(ret[f.name], f.resolvedType);
      }
    });
  
    return ret;
  };
  return encode
}

export const finalize = (formState: unknown, type: protobuf.Type, context: AutoFormContext) => {
  const protoObj = formToProtoObj(context)(formState, type) as Record<string, unknown>;
  const prunedProtoObj = pruneUnselectedOneofValues(protoObj, type);

  return prunedProtoObj;
};

export const protoObjToForm = (protoObj: unknown, type: protobuf.Type | protobuf.Enum | null) => {
  if (!(type instanceof protobuf.Type)) return protoObj;
  const ret = { ...(protoObj as Record<string, unknown>) };

  type.fieldsArray.forEach((f) => {
    if (!ret[f.name]) return;

    if (f.repeated) {
      ret[f.name] = (ret[f.name] as unknown[]).map((value) => ({ $value: protoObjToForm(value, f.resolvedType) }));
    } else if (f.map) {
      ret[f.name] = Object.entries(ret[f.name] as Record<string, unknown>).map(([$key, $value]) => ({
        $key,
        $value: protoObjToForm($value, f.resolvedType),
      }));
    } else {
      ret[f.name] = protoObjToForm(ret[f.name], f.resolvedType);
    }
  });

  return ret;
};
