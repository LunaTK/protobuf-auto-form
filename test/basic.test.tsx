import protobuf from 'protobufjs';
import React from 'react';
import renderer from 'react-test-renderer';
import {
  describe, expect, it, test,
} from 'vitest';
import descriptor from '../example/proto.json';
import AutoForm from '../src/AutoForm';

const namespace = protobuf.Namespace.fromJSON('', descriptor);
const { Field } = AutoForm;

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();
  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);
  return result as renderer.ReactTestRendererJSON;
}

test('Snapshot test', () => {
  const component = renderer.create(
    <div>hi</div>,
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});

describe('rendering', () => {
  it('basic form rendering', () => {
    const component = renderer.create(
      <AutoForm namespace={namespace} messageType="User" />,
    );
    expect(component).toMatchSnapshot();
  });

  it('field name override', () => {
    const component = renderer.create(
      <AutoForm namespace={namespace} messageType="User">
        <Field name="name" label="User Name" />
      </AutoForm>,
    );
    expect(component).toMatchSnapshot();
  });
});
