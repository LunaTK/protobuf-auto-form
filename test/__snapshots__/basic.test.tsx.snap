// Vitest Snapshot v1

exports[`Link changes the class when hovered 1`] = `
<div>
  hi
</div>
`;

exports[`Link changes the class when hovered 2`] = `
<div>
  hi
</div>
`;

exports[`Snapshot test 1`] = `
<div>
  hi
</div>
`;

exports[`Snapshot test 2`] = `
<div>
  hi
</div>
`;
